### ISSUE TITLE
Intermittent compile crashes with "Errno::EINVAL: Invalid argument @ io_write"

### ISSUE THREAD

**Issue by sarahg** [nanoc/nanoc#1635]
### Steps to reproduce

Can't reproduce this regularly yet, but 3 of us on the [GitLab Docs](https://gitlab.com/gitlab-org/gitlab-docs) project have run into this error multiple times when compiling locally:

```
      create  [0.18s]  public/assets/stylesheets/toc.css

Captain! We've been hit!

Errno::EINVAL: Invalid argument @ io_write - /Users/eread/Projects/gitlab-docs/tmp/nanoc/20ba532647207/compiled_content.data.db

  0. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/store.rb:132:in `binwrite'
  1. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/store.rb:132:in `write_obj_to_file'
  2. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/store.rb:111:in `store_uninstrumented'
  3. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/store.rb:90:in `block in store'
  4. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/instrumentor.rb:27:in `call'
  5. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/store.rb:89:in `store'
  6. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/compiled_content_cache.rb:64:in `block in store'
  7. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/compiled_content_cache.rb:64:in `each'
  8. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/compiled_content_cache.rb:64:in `store'
  9. /Users/eread/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/nanoc-core-4.12.12/lib/nanoc/core/compilation_stages/compile_reps.rb:37:in `run'
```

This comes up when running `bundle exec nanoc compile` and `bundle exec nanoc view`.

We've seen this with Ruby 2.7.6 and 2.7.7, and nanoc 4.12.9 and 4.12.12.

### Expected behavior

Nanoc does not crash when compiling.

### Actual behavior

Nanoc fails to compile.

### Details

- Opening a new terminal and running compile again seems to clear it up.
- Happens at the very end of the compile process.

### Crash log

Couple other examples:

- https://gist.github.com/sarahg/e4197f9f2aef8c90f45a4bea226b3774

![image](https://user-images.githubusercontent.com/667501/205144419-7ce11c56-e389-48a7-83a8-ca749134b4bd.png)

**Comment by denisdefreyne** [comment]
This seems to happen when the data argument to `File.write(…)` is too large. Something to reproduce it with:

```ruby
require 'stringio'

io = StringIO.new
100000.times {
  io << 'x' * 100000
}

string = io.string
File.write('tmp.dat', string)
```

```
blah.rb:9:in `write': Invalid argument @ io_write - tmp.dat (Errno::EINVAL)
	from blah.rb:9:in `<main>'
```

**Comment by denisdefreyne** [comment]
@sarahg Can you check whether #1636 fixes the issue?

**Comment by sarahg** [comment]
@denisdefreyne I'm running nanoc with #1636 this afternoon while doing some dev work on the site. No immediate issues, but I was not able to consistently reproduce it. I'll ask some other folks who have seen this to also try the fix. Thanks for taking a look!

**Comment by denisdefreyne** [comment]
A fix for this is available in Nanoc 4.12.14.
